### DISCUSSION TITLE
How to filter warnings with --pyargs option and pyproject.toml

### DISCUSSION THREAD

**Discussion by skoudoro** [pytest-dev/pytest#12152] [General] [answered]
Hi All, Hi Pytest team,

I tried to filter some warnings but it seems not to work. I would like to know if there is a way to filter those warnings and understand what I am doing wrong.

I use the following command :

```python
# install the project to the environment
pip install  dipy
# create a new folder
mkdir for_testing
# leave the source folder and go to this new empty folder
cd for_testing
# Copy the pyproject.toml  to use it as a configuration files
cp ../pyproject.toml .
# Run the tests
pytest -c pyproject.toml -svv --doctest-modules --verbose --durations=10 --pyargs dipy
```

my pyproject.toml contains the following information:
```
[tool.pytest.ini_options]
addopts = "--ignore=dipy/testing/decorators.py --ignore-glob='bench*.py' --ignore-glob=**/benchmarks/*"
filterwarnings = [
    'ignore:.*You do not have FURY installed.*:UserWarning',
]
```

I am ok if there is an alternative without the `pyproject.toml`

Thank you for your feedback

PS : I have search on the repo, but could not find an associated issue.

**Comment by skoudoro** [comment]
Hi all, 

I will be happy to clarify if there is something not clear in my question.

Any ideas ? feedback ?

Thanks

**Reply by nicoddemus** [reply]
Hey @skoudoro,

`pyproject.toml` is fine, unfortunately some warnings cannot be filtered out by pytest easily due to the internal design. Can you show the warnings summary you get?

**Reply by jhlegarreta** [reply]
Hi @nicoddemus, chiming in here as I contribute to DIPY and as we are only this warning away from being able to turn warnings into errors in testing to avoid accumulating other warnings over time.

The issue is that in DIPY there is an optional dependency for the FURY package: https://github.com/dipy/dipy/blob/master/pyproject.toml#L111
https://github.com/dipy/dipy/blob/master/pyproject.toml#L92

So in the `__init__.py` file of the `viz` module, we have a [statement that raises a warning](https://github.com/dipy/dipy/blob/master/dipy/viz/__init__.py#L19) if that dependency is not found. So when running the tests without that dependency installed, we systematically get the warning e.g. https://github.com/dipy/dipy/actions/runs/8801041490/job/24153517109#step:10:4397:
```
=============================== warnings summary ===============================
../venv/lib/python3.10/site-packages/dipy/viz/__init__.py:20
  /home/runner/work/dipy/dipy/venv/lib/python3.10/site-packages/dipy/viz/__init__.py:20: UserWarning: You do not have FURY installed. Therefore, 3D visualization functions will not work for you. Please install or upgrade FURY using pip install -U furyFor detailed installation instructions visit: https://fury.gl/
    warnings.warn(
```

We are willing to tell `pytest` e.g. https://github.com/dipy/dipy/blob/master/pyproject.toml#L184 to filter that warning explicitly so that we can turn warnings into errors, as mentioned above.

I understand that this is different from filtering warnings in a test method, but we would like to know the proper way to filter the mentioned warning.

**Reply by jhlegarreta** [reply, marked as answer]
@nicoddemus Looks like we found out why the warning was being raised: this commits fixes it:
https://github.com/dipy/dipy/pull/3158/commits/79e64a8f25f3cec403708442ead9be29fa884a3f

Unless @skoudoro has some other comment, the issue may be closed.

Thanks for developing `pytest` and providing support.

**Reply by skoudoro** [reply]
Yes, it can be closed, thanks @nicoddemus and @jhlegarreta for the help
